src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs

```diff
@@ -90,10 +116,49 @@ private static void HandleEndObject(ref ReadStack state)
             }
             else
             {
+                // Set directly when handling preserved array
+                bool setPropertyDirectly = state.Current.IsPreservedArray;
                 state.Pop();
 
-                ApplyObjectToEnumerable(value, ref state);
+                ApplyObjectToEnumerable(value, ref state, setPropertyDirectly);
             }
         }
+
+        private static object GetPreservedArrayValue(ref ReadStack state)
+        {
+            // Preserved JSON arrays are wrapped into JsonPreservedReference<T> where T is the original type of the enumerable
+            // and Values is the actual enumerable instance being preserved.
+            JsonPropertyInfo info = state.Current.JsonClassInfo.PropertyCache["Values"];
+            object value = info.GetValueAsObject(state.Current.ReturnValue);
+
+            if (value == null)
+            {
+                ThrowHelper.ThrowJsonException_MetadataPreservedArrayValuesNotFound(info.DeclaredPropertyType);
+            }
+
+            return value;
+        }
+
+        private static void HandlePreservedArray(ref ReadStack state, JsonSerializerOptions options)
+        {
+            // Check we are not parsing into immutable or array.
+            if (state.Current.JsonPropertyInfo.EnumerableConverter != null)
+            {
+                ThrowHelper.ThrowJsonException_MetadataCannotParsePreservedObjectIntoImmutable(state.Current.JsonPropertyInfo.DeclaredPropertyType);
```

Comment:
Actually, this is a good point. Newtonsoft.Json Serializes Immutable types with metadata, but on Deserialize, it throws if this metadata is present on those types... therefore an immutable type is unable to roundtrip without failing on deserialization when using Newtonsoft's.

Since we already choose to apply the rule of not emitting metadata for value types (structs) on Serialization, I lean towards extending such rule to Immutable types and therefore keep this validation on Deserialize.

Reply:
Let's track this in https://github.com/dotnet/runtime/issues/1881.